src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8389,40 +8472,182 @@ protected MethodGroupResolution BindExtensionMethod(
                                       OverloadResolution.Options.DynamicConvertsToAnything)) == 0);
 
             var firstResult = new MethodGroupResolution();
-            AnalyzedArguments actualArguments = null;
+            AnalyzedArguments? actualArguments = null;
+
+            foreach (var scope in new ExtensionScopes(this))
+            {
+                // PROTOTYPE we are temporarily prioritizing new extension members over classic extension methods, instead of mixing them
+                // PROTOTYPE we'll want to allow resolution of extension members on values of type parameters (but not on type parameter types)
+                if (!left.Type.IsTypeParameter()
+                    && tryResolveExtensionMember(this, expression, memberName, analyzedArguments, left, typeArgumentsWithAnnotations,
+                        options, returnRefKind, in callingConvention, returnType, withDependencies, scope,
+                        out MethodGroupResolution extensionResult))
+                {
+                    if (extensionResult.IsNonMethodExtensionMember(out _))
+                    {
+                        return extensionResult;
+                    }
+
+                    // PROTOTYPE(instance) we'll want to revisit to merge extension methods and methods from extension types within the same scope
+
+                    // If the search in the current scope resulted in any applicable method from an extension type
+                    // (regardless of whether a best applicable method could be determined) then our search is complete.
+                    // Otherwise, store aside the first non-applicable result and continue searching for an applicable result.
+                    if (extensionResult.HasAnyApplicableMethod)
+                    {
+                        firstResult.Free();
+                        return extensionResult;
+                    }
+                    else if (firstResult.IsEmpty)
+                    {
+                        firstResult = extensionResult;
```

Comment:
>firstResult = extensionResult; [](http://example.com/codeflow?start=24&length=30)

We don't need to free ```firstResult``` here?

Reply:
`firstResult.IsEmpty` indicates that don't have anything to free (in particular, no analyzed arguments). I'll add an assertion to clarify.